analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Checks/InvocationCheck.cs

```diff
@@ -26,27 +29,49 @@ namespace SonarAnalyzer.SymbolicExecution.Roslyn.Checks
 {
     internal class InvocationCheck : SymbolicCheck
     {
+        private static ImmutableArray<KnownType> throwHelperAttributes = new[]
+        {
+            KnownType.System_Diagnostics_CodeAnalysis_DoesNotReturnAttribute,
+            KnownType.JetBrains_Annotations_TerminatesProgramAttribute,
+        }.ToImmutableArray();
+
         public override ProgramState[] PreProcess(SymbolicContext context) =>
             context.Operation.Instance.Kind == OperationKindEx.Invocation
             && PreProcess(context, IInvocationOperationWrapper.FromOperation(context.Operation.Instance)) is { } newState
                 ? newState
                 : base.PreProcess(context);
 
-        private static ProgramState[] PreProcess(SymbolicContext context, IInvocationOperationWrapper invocation) =>
-            invocation.TargetMethod.IsAny(KnownType.System_String, nameof(string.IsNullOrEmpty), nameof(string.IsNullOrWhiteSpace))
-                ? ProcessStringIsNullOrEmpty(context, invocation)
-                : null;
+        private static ProgramState[] PreProcess(SymbolicContext context, IInvocationOperationWrapper invocation)
+        {
+            if (invocation.TargetMethod.IsAny(KnownType.System_String, nameof(string.IsNullOrEmpty), nameof(string.IsNullOrWhiteSpace)))
+            {
+                return ProcessStringIsNullOrEmpty(context.State, invocation);
+            }
+            else if (IsThrowHelper(invocation))
+            {
+                return EmptyStates; // Stop processing
+            }
+            return null;
+        }
+
+        private static bool IsThrowHelper(IInvocationOperationWrapper invocation) =>
+            KnownMethods.IsDebugFail(invocation.TargetMethod)
+            || invocation.TargetMethod.IsAny(KnownType.System_Environment, nameof(Environment.FailFast), nameof(Environment.Exit))
+            || invocation.TargetMethod.GetAttributes(throwHelperAttributes).Any();
 
-        private static ProgramState[] ProcessStringIsNullOrEmpty(SymbolicContext context, IInvocationOperationWrapper invocation) =>
-            context.State[invocation.Arguments[0].ToArgument().Value]?.Constraint<ObjectConstraint>() switch
+        private static ProgramState[] ProcessStringIsNullOrEmpty(ProgramState state, IInvocationOperationWrapper invocation) =>
```

Comment:
Why this change? `context` was handy and this made the implementation header to read. Should be reverted

Reply:
I don't like the idea to pass `SymbolicContext` as soon as there is more than one thing going on. You want to be able to do `state = DoSomeMore(state)` at one point and reorder it. It is hard to do if you pass `SymbolicContext`.

Reply:
Content of this method is pretty much done in long run. The change made it harder to read.